crates/bls12381/src/fields/fp.rs

```diff
@@ -0,0 +1,504 @@
+use bellpepper_core::{
+    boolean::{AllocatedBit, Boolean},
+    ConstraintSystem, SynthesisError,
+};
+use bellpepper_emulated::field_element::{
+    EmulatedFieldElement, EmulatedFieldParams, PseudoMersennePrime,
+};
+use bls12_381::fp::Fp as BlsFp;
+use ff::{PrimeField, PrimeFieldBits};
+use num_bigint::{BigInt, Sign};
+
+#[derive(Debug)]
+pub struct Bls12381FpParams;
+
+impl EmulatedFieldParams for Bls12381FpParams {
+    // TODO: Depending on the native field, different limb/bit pairs are more optimal and have less waste. This should be customizable and not hardcoded
+    // for example, in the pasta field, 4/96 could be used instead
+    fn num_limbs() -> usize {
+        6
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fp<F> = EmulatedFieldElement<F, Bls12381FpParams>;
+
+#[derive(Clone)]
+pub struct AllocatedFieldElement<F: PrimeField + PrimeFieldBits>(pub(crate) Bls12381Fp<F>);
```

Comment:
This is a weird name, since data of this type will *not* be allocated. I think it would be better to follow the naming convention that reserves this prefix for the allocated version of some 'vanilla' data structure.

Reply:
I agree -- I am not a fan of all the `Allocated<whatever>` pattern I ended up sticking with. I will drop the `Allocated` prefix from all the main types used